### PR TITLE
feat: Extend DelayedScalers into arbitary UpdatingScalers

### PR DESCRIPTION
**Pull request by HCookie**
## Description
Removes `DelayedScalers` and add `UpdatingScalers`. 
Allows for more fine control as to when and how the `scalers` are updated.

## What problem does this change solve?
Part of #370 

##  Additional notes ##

***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***

### CODE REVIEW

File: training/src/anemoi/training/losses/scalers/loss_weights_mask.py [hunk 1]
```diff
@@ -23,25 +23,11 @@
 LOGGER = logging.getLogger(__name__)
 
 
-class NaNMaskScaler(BaseDelayedScaler):
+class NaNMaskScaler(BaseUpdatingScaler):
 
     scale_dims: tuple[TensorDim] = (TensorDim.GRID, TensorDim.VARIABLE)
 
-    def __init__(self, norm: str | None = None, **kwargs) -> None:
```
Comment (sahahner) [inline training/src/anemoi/training/losses/scalers/loss_weights_mask.py:26]: I have tested this PR in training: If this init function is removed, the `__init__` function of the `base_scaler` should allow `**kwargs` to support this scaler's initialisation.

**Comment by sahahner** [comment]
Thank you very much for this PR. I like the solution with the callbacks as it is flexible and can be reused for other scalers.

I see you have tried to cover many different cases. Nevertheless, the NaN_weights need to be updated after `y_pred = self(x)` (after the forward pass, before the training loss calculation). None of the provided callbacks cover that case. 
Would it be possible to include that case, please? It seems to be non-standard callback event, so we might need to come up with a name. What do you think about `on_after_forward`? (pytorch lightning for example support `on_after_backward`).

Nevertheless, I do feel that this PR increases the complexity of the forecaster by calling a few more functions without any added functionality. We might want to remove the callbacks that are actually not used and leave it for future PRs that require this functionality. Either way, I might not be aware of other use cases that are planned to be included.



Thank you very much for this PR — I like the callback-based solution, as it’s flexible and potentially reusable for other scalers.

I see you've tried to cover many cases, which is great. However, for https://github.com/ecmwf/anemoi-core/issues/370 NaN_weights need to be updated after the forward pass (`y_pred = self(x)`) and before the training loss is computed. None of the current callbacks seem to handle this case. Would it be possible to add a callback event for this? Since it’s not a standard one, we might need to define a new name — what do you think about `on_after_forward`? (Similar to how PyTorch Lightning uses `on_after_backward`.)

That said, I feel this PR adds some complexity by introducing multiple callbacks that aren’t currently used. We might want to consider removing the unused ones for now and adding them later when there’s a clear need. Of course, there might be future use cases in mind that I do not know about.

For context, I was able to build https://github.com/ecmwf/anemoi-core/pull/378 on top of this for now after adjusting the position of `ON_TRAIN_BATCH_END`.

File: training/src/anemoi/training/losses/scalers/loss_weights_mask.py (hunk 1 above)
Comment (HCookie) [reply training/src/anemoi/training/losses/scalers/loss_weights_mask.py:26]: What for? What kwargs are passed?

File: training/src/anemoi/training/losses/scalers/base_scaler.py [hunk 2]
```diff
@@ -84,19 +94,121 @@ def get_scaling(self) -> SCALER_DTYPE:
...
+            Scaling values as a numpy array.
+        """
+        if self._cached_scaling_values is not None:
+            LOGGER.debug("Using cached scaling values for %s.", self.__class__.__name__)
+            return self._cached_scaling_values
+
+        return self.initial_scaling_values()
+
+    def get_scaling(self) -> SCALER_DTYPE:
+        """Get scaling values based on the initial scaling values callback."""
```
Comment (anaprietonem) [inline training/src/anemoi/training/losses/scalers/base_scaler.py:131]: What's the motivation for the caching of the scaling values?

File: training/src/anemoi/training/losses/scalers/base_scaler.py [hunk 3]
```diff
@@ -84,19 +94,121 @@ def get_scaling(self) -> SCALER_DTYPE:
         return scale_dims, scaler_values
 
 
-class BaseDelayedScaler(BaseScaler):
-    """Base class for delayed Scalers.
+class AvailableCallbacks(StrEnum):
+    INITIAL_SCALING_VALUES = "initial_scaling_values"
```
Comment (anaprietonem) [inline training/src/anemoi/training/losses/scalers/base_scaler.py]: To my understanding scalers take default values and depending on the config values those can change for certain grid point, variables or vertical levels. What we are thinking now is those also might change according to certain requirements imposed by pre-processors like the imputers. I guess then I struggle to see the need for this hooks and how do they connect with the pre-processors

Comment (anaprietonem) [reply training/src/anemoi/training/losses/scalers/base_scaler.py]: Okey @sahahner kindly explained the fact that we need updates on training and validation as the mask has the batch dimension and those can change. I guess simplifying it so we just use the hooks we need as @sahahner  mentioned above could be good, but I wonder is there is no way to do directly on the forecaster? On the rollout_step directly . So directly doing what the `NaNMaskScaler` is doing on `on_training_start` on the forecaster method to update the scalers. That's method is called on the rollout_step so the update would happen both for training and validation?

Comment (HCookie) [reply training/src/anemoi/training/losses/scalers/base_scaler.py]: I tried to keep fairly general to capture future use cases. I'm happy to remove the superfolous callbacks for now and just keep start, and rollout_step ones

File: training/src/anemoi/training/losses/scalers/base_scaler.py (hunk 2 above)
Comment (HCookie) [reply training/src/anemoi/training/losses/scalers/base_scaler.py:131]: They will be called for the get on each callback, but only some will update them, hence caching

File: training/src/anemoi/training/losses/scalers/base_scaler.py (hunk 3 above)
Comment (anaprietonem) [reply training/src/anemoi/training/losses/scalers/base_scaler.py]: I appreciate the nice work you're doing here and that the callback is smart trick. But I think the imputer/handling of the nans is a special case so I'd be in favour of reducing complexity as much as possible so in the future we can identify the changes of https://github.com/ecmwf/anemoi-core/issues/370 much easier.  A few things from the current design that are still unclear to me are:
- the `get_scaling_values` of the BaseUpdatingScaler or any class that inherits from it - where is that used?
- We discussed that the scalers need to be updated per batch for training and validation but the NaNMaskScaler just has the 'on_training_start' implemented. 
- In https://github.com/ecmwf/anemoi-core/pull/378 - the pre_processor.loss_mask_training is being recomputed for each batch under the transform that gets called in the forecaster rollout_step. So what I was wondering is why it would be enough to call the update method there two?
```
        for pre_processor in model.pre_processors.processors.values():
            if hasattr(pre_processor, "loss_mask_training"):
                loss_weights_mask = loss_weights_mask * pre_processor.loss_mask_training.cpu().numpy()
```
since the result of `pre_processor.loss_mask_training` would be already taking the right shape depending if it's training or validation, so then we could simplify the logic

Comment (sahahner) [reply training/src/anemoi/training/losses/scalers/base_scaler.py]: > * We discussed that the scalers need to be updated per batch for training and validation but the NaNMaskScaler just has the 'on_training_start' implemented.

This change is implemented in #378 where we implement the logic of the nan calculation of the fly.

Comment (anaprietonem) [reply training/src/anemoi/training/losses/scalers/base_scaler.py]: Yes, maybe my comment was not clear. What I mean is looking at 378 and the changes at 371 - I think there is a mismatch in the logic - because of callback for update is 'on_training_start' and the update of the nan mask is done on the rollout step per batch.  But this also to me, means we could simply some of the logic here.

File: training/src/anemoi/training/losses/scalers/base_scaler.py (hunk 2 above)
Comment (sahahner) [reply training/src/anemoi/training/losses/scalers/base_scaler.py:131]: I see the motivation behind the caching, but it seems to be causing me errors as these cached scaling values are not on the device. Could we access the buffered scalers here?

File: training/src/anemoi/training/losses/scalers/loss_weights_mask.py (hunk 1 above)
Comment (sahahner) [reply training/src/anemoi/training/losses/scalers/loss_weights_mask.py:26]: There was an error in my config.

Comment (sahahner) [reply training/src/anemoi/training/losses/scalers/loss_weights_mask.py:26]: I had another look and indeed this line seems to pass `group_config` to the updating scaler: https://github.com/ecmwf/anemoi-core/blob/main/training/src/anemoi/training/losses/scalers/scalers.py#L30. The other methods override the __init__ which is why i think just the nan_mask fails.

File: training/src/anemoi/training/losses/scalers/base_scaler.py (hunk 3 above)
Comment (sahahner) [reply training/src/anemoi/training/losses/scalers/base_scaler.py]: I am not sure I understand the comment. `on_training_start` will actually not be implemented anymore for the loss mask.

File: training/src/anemoi/training/losses/scalers/base_scaler.py [hunk 4]
```diff
@@ -84,19 +94,121 @@ def get_scaling(self) -> SCALER_DTYPE:
...
+    """Base class for updating scalers.
 
-    The delayed scalers are only initialised when creating all the scalers, but its value is
-    computed during the first iteration of the training loop. This delayed scalers are suitable
-    for scalers requiring information from the `model.pre_processors`.
+    The updating scalers have a variety of callback methods associated with them,
+    which are called during the training loop. These methods allow the scalers to
+    update their values based on the current state of the model and the training data.
+
+    The callback methods are expected to return a np.ndarray of scaling values,
```
Comment (sahahner) [inline training/src/anemoi/training/losses/scalers/base_scaler.py:105]: what if output is the default which is None? -> no update on scaler

File: training/src/anemoi/training/losses/scalers/base_scaler.py (hunk 2 above)
Comment (HCookie) [reply training/src/anemoi/training/losses/scalers/base_scaler.py:131]: Caching has been removed

File: training/src/anemoi/training/train/forecaster/forecaster.py [hunk 5]
```diff
@@ -234,11 +235,19 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
     def on_load_checkpoint(self, checkpoint: torch.nn.module) -> None:
         self._ckpt_model_name_to_index = checkpoint["hyper_parameters"]["data_indices"].name_to_index
 
-    def define_delayed_scalers(self) -> None:
+    def update_scalers(self, callback: AvailableCallbacks) -> None:
         """Update delayed scalers such as the loss weights mask for imputed variables."""
```
Comment (anaprietonem) [inline training/src/anemoi/training/train/forecaster/forecaster.py]: docstring update - no longer delayed?

File: training/src/anemoi/training/train/forecaster/forecaster.py [hunk 6]
```diff
@@ -234,11 +235,19 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
-        for name, scaler_builder in self.delayed_scaler_builders.items():
-            self.scalers[name] = scaler_builder.get_delayed_scaling(model=self.model)
-            self.loss.update_scaler(scaler=self.scalers[name][1], name=name)
+        for name, scaler_builder in self.updating_scalars.items():
+            scaler = scaler_builder.update_scaling_values(callback, model=self.model)
+            if scaler is None:
+                continue
+
```
Comment (anaprietonem) [inline training/src/anemoi/training/train/forecaster/forecaster.py]: why scalers[1]? could be good to add some comments about what is going on inside each loop

File: training/src/anemoi/training/train/forecaster/forecaster.py (hunk 5 above)
Comment (HCookie) [inline training/src/anemoi/training/train/forecaster/forecaster.py]:
```suggestion
        """Update scalers, calling the defined function on them, updating if not None"""
```

File: training/src/anemoi/training/train/forecaster/forecaster.py [hunk 7]
```diff
@@ -234,11 +235,19 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            if name in self.loss.scaler:
+                self.loss.update_scaler(scaler=scaler[1], name=name)
+
+            for metric in self.metrics.values():
+                if name in metric.scaler:
+                    metric.update_scaler(scaler=scaler[1], name=name)
```
Comment (HCookie) [inline training/src/anemoi/training/train/forecaster/forecaster.py]:
```suggestion
        for name, scaler_builder in self.updating_scalars.items():
            scaler = scaler_builder.update_scaling_values(callback, model=self.model)
            if scaler is None: # If scalar is None, no update to be applied
                continue

            if name in self.loss.scaler: # If scalar in loss, update it
                self.loss.update_scaler(scaler=scaler[1], name=name) # Only update the values

            for metric in self.metrics.values(): # If scalar in metrics, update it
                if name in metric.scaler:
                    metric.update_scaler(scaler=scaler[1], name=name) # Only update the values
```

**Review by HCookie — COMMENTED** [review]
@anaprietonem Thoughts?

File: training/src/anemoi/training/train/forecaster/interpolator.py [hunk 8]
```diff
@@ -99,8 +100,15 @@ def _step(
 
         # Delayed scalers need to be initialized after the pre-processors once
         if self.is_first_step:
-            self.define_delayed_scalers()
+            self.update_scalars(callback=AvailableCallbacks.ON_TRAINING_START)
             self.is_first_step = False
+        self.update_scalars(
+            callback=(
```
Comment (anaprietonem) [inline training/src/anemoi/training/train/forecaster/interpolator.py]: I think we thought that the valid/training abstraction was not needed, and just keep the training start and batch one. So this would need update?

File: training/src/anemoi/training/train/forecaster/forecaster.py [hunk 9]
```diff
@@ -412,6 +423,10 @@ def rollout_step(
 
             yield loss, metrics_next, y_pred
 
+        self.update_scalers(
```
Comment (anaprietonem) [inline training/src/anemoi/training/train/forecaster/forecaster.py]: why are we updating the scalar after the rollout step again?

File: training/src/anemoi/training/losses/scalers/base_scaler.py [hunk 10]
```diff
@@ -84,19 +94,87 @@ def get_scaling(self) -> SCALER_DTYPE:
         return scale_dims, scaler_values
 
 
-class BaseDelayedScaler(BaseScaler):
-    """Base class for delayed Scalers.
+class AvailableCallbacks(StrEnum):
+    INITIAL_SCALING_VALUES = "initial_scaling_values"
```
Comment (anaprietonem) [inline training/src/anemoi/training/losses/scalers/base_scaler.py]: why we we have 'INITIAL_SCALING_VALUES'  as part of the callback?

File: training/src/anemoi/training/train/forecaster/forecaster.py [hunk 11]
```diff
@@ -412,6 +423,10 @@ def rollout_step(
...
+            callback=(AvailableCallbacks.ON_BATCH_START),
+        )
```
Comment (HCookie) [inline training/src/anemoi/training/train/forecaster/forecaster.py]:
```suggestion
```

File: training/src/anemoi/training/losses/scalers/base_scaler.py (hunk 10 above)
Comment (HCookie) [inline training/src/anemoi/training/losses/scalers/base_scaler.py]:
```suggestion
```

File: training/src/anemoi/training/train/forecaster/ensforecaster.py [hunk 12]
```diff
@@ -238,8 +239,15 @@ def rollout_step(
...
-            self.define_delayed_scalers()
+            self.update_scalers(callback=AvailableCallbacks.ON_TRAINING_START)
             self.is_first_step = False
+        self.update_scalers(
+            callback=(
+                AvailableCallbacks.ON_TRAIN_BATCH_START
+                if not validation_mode
+                else AvailableCallbacks.ON_VALID_BATCH_START
+            ),
+        )
```
Comment (HCookie) [inline training/src/anemoi/training/train/forecaster/ensforecaster.py]:
```suggestion
        self.update_scalers(callback=AvailableCallbacks.ON_BATCH_START)
```

File: training/src/anemoi/training/train/forecaster/interpolator.py [hunk 13]
```diff
@@ -99,8 +100,15 @@ def _step(
...
+                AvailableCallbacks.ON_TRAIN_BATCH_START
+                if not validation_mode
+                else AvailableCallbacks.ON_VALID_BATCH_START
+            ),
+        )
```
Comment (HCookie) [inline training/src/anemoi/training/train/forecaster/interpolator.py]:
```suggestion
        self.update_scalers(callback=AvailableCallbacks.ON_BATCH_START)
```

File: training/src/anemoi/training/train/forecaster/forecaster.py (hunk 9 above)
Comment (HCookie) [reply training/src/anemoi/training/train/forecaster/forecaster.py]: Removed

File: training/src/anemoi/training/train/forecaster/interpolator.py (hunk 8 above)
Comment (HCookie) [reply training/src/anemoi/training/train/forecaster/interpolator.py]: Corrected

**Review by anaprietonem — COMMENTED** [review]
This PR looks good. @HCookie addressed the comments and I think the logic is now simplified, keeping a trade-off between flexibility and complexity. I think before merging it would still need to be documented the new logic for updating scalars (could use the imputer as an example). 

I think the new logic to use None when no update is done it's also neat, and could open the door to simplify further the scalers abstraction (ie baseScalars could also be seen as 'updating' ones with None updates). This could be revisit in the future

File: training/docs/modules/losses.rst [hunk 14]
```diff
@@ -203,6 +203,49 @@ If multiple groups are defined for a variable, the first group in the
 `variable_groups` is used. If the variable is not in any group, it is
 assigned to the default group.
 
+Custom Scalars
```
Comment (anaprietonem) [inline training/docs/modules/losses.rst:206]: Very nice!

**Review by anaprietonem — DISMISSED** [review]
Good work @HCookie and nice docs!

**Review by anaprietonem — APPROVED** [review]
Changes were addressed and tests are passing! good work Harrison, good to be merged!
